src/handlers/web3.ts

```diff
@@ -0,0 +1,679 @@
+  transaction: Pick<
+    NewTransactionNonNullable,
+    'asset' | 'from' | 'to' | 'amount' | 'gasPrice' | 'gasLimit'
+  >
+): Promise<TransactionDetailsReturned> => {
+  const value = convertAmountToRawAmount(
+    transaction.amount,
+    transaction.asset.decimals
+  );
+  const recipient = (await resolveNameOrAddress(transaction.to)) as string;
+  const data = getDataForTokenTransfer(value, recipient);
+  return {
+    data,
+    from: transaction.from,
+    gasLimit: transaction.gasLimit,
+    gasPrice: transaction.gasPrice,
+    to: transaction.asset.address,
+  };
+};
+
+/**
+ * @desc Transforms a new transaction into signable transaction.
+ * @param transaction The new transaction.
+ * @return The transaction details.
+ */
+export const createSignableTransaction = async (
+  transaction: NewTransactionNonNullable
+): Promise<TransactionDetailsReturned> => {
+  if (
+    get(transaction, 'asset.address') === ETH_ADDRESS ||
+    get(transaction, 'asset.address') === ARBITRUM_ETH_ADDRESS ||
+    get(transaction, 'asset.address') === OPTIMISM_ETH_ADDRESS ||
+    get(transaction, 'asset.address') === MATIC_POLYGON_ADDRESS
+  ) {
+    return getTxDetails(transaction);
+  }
+  const isNft = get(transaction, 'asset.type') === AssetType.nft;
+  const result = isNft
+    ? await getTransferNftTransaction(transaction)
+    : await getTransferTokenTransaction(transaction);
+  return getTxDetails(result);
+};
+
+/**
+ * @desc Estimates the balance portion for a given asset.
+ * @param asset The asset to check.
+ * @returns The estimated portion.
+ */
+const estimateAssetBalancePortion = (asset: ParsedAddressAsset): string => {
+  if (!(asset.type === AssetType.nft)) {
+    const assetBalance = get(asset, 'balance.amount');
+    const decimals = get(asset, 'decimals');
+    const portion = multiply(assetBalance, 0.1);
+    const trimmed = handleSignificantDecimals(portion, decimals);
+    return convertAmountToRawAmount(trimmed, decimals);
+  }
+  return '0';
+};
+
+/**
+ * @desc Generates a transaction data string for a token transfer.
+ * @param value The value to transfer.
+ * @param to The recipient address.
+ * @returns The data string for the transaction.
+ */
+export const getDataForTokenTransfer = (value: string, to: string): string => {
+  const transferMethodHash = smartContractMethods.token_transfer.hash;
+  const data = ethereumUtils.getDataString(transferMethodHash, [
+    ethereumUtils.removeHexPrefix(to),
+    convertStringToHex(value),
+  ]);
+  return data;
+};
+
+/**
+ * @desc Returns a transaction data string for an NFT transfer.
+ * @param from The sender's address.
+ * @param to The recipient's address.
+ * @param asset The asset to transfer.
+ * @returns The data string.
+ */
+export const getDataForNftTransfer = (
+  from: string,
+  to: string,
+  asset: ParsedAddressAsset
+): string => {
+  const nftVersion = get(asset, 'asset_contract.nft_version');
+  const schema_name = get(asset, 'asset_contract.schema_name');
```

Comment:
The fields requested on this line and the line above it don't exist on the `ParsedAddressAsset` type. Is the type incomplete, or is this a bug in the sense that they _always_ resolve to `undefined`?

Side note, probably worth a chat — this is a good example of why it might be a good idea to avoid `get` as we migrate to TypeScript because it masks type errors like this.

Reply:
Yes, fixed this and changed it to remove `get`, thanks! Those fields are sourced from the OpenSea API, so I added them as optional fields to the asset definition.